File: src/reasoners/reflexion.py
from __future__ import annotations

from typing import Optional

from .base import Reasoner, ReasoningTrace


class ReflexionReasoner(Reasoner):
    def solve(self, query: str, initial_solution: Optional[str] = None) -> ReasoningTrace:
        max_trials = self.config.get("max_trials", 3)

        attempt = initial_solution or ""
        reflections: list[str] = []

        for _ in range(max_trials):
            if not attempt:
                attempt = self._chat(
                    [
                        {
                            "role": "user",
                            "content": query
                            + "\nThink step by step. End with 'Final answer: ...'.",
                        }
                    ]
                )

            critique = self._chat(
                [
                    {
                        "role": "user",
                        "content": f"""
[Problem]
{query}

[Attempt]
{attempt}

Critique weaknesses or errors.
""",
                    }
                ]
            )
            reflections.append(critique)

            attempt = self._chat(
                [
                    {
                        "role": "user",
                        "content": f"""
[Problem]
{query}

Reflections:
{chr(10).join(reflections)}

Rewrite solution based on reflections.
End with 'Final answer: ...'.
""",
                    }
                ]
            )

        reasoning, answer = self._split(attempt)
        return ReasoningTrace(
            reasoning,
            answer,
            {"method": "reflexion", "reflections": reflections},
        )


__all__ = ["ReflexionReasoner"]

File: tui/src/reasoners/reflexion.ts
// ── Reflexion ──────────────────────────────────────────────────────────────────
// Adapted from Python src/reasoners/reflexion.py.
//
// Iterate: attempt → critique → rewrite based on reflections.

import { BaseReasoner, type ReasoningTrace } from "./base.js";

interface ReflexionConfig {
	maxTrials?: number;
	temperature?: number;
	maxTokens?: number;
	[key: string]: unknown;
}

export class ReflexionReasoner extends BaseReasoner {
	config: ReflexionConfig;

	constructor(
		llm: import("../agent-core/backend.js").LLMBackend,
		config: ReflexionConfig = {},
	) {
		super(llm, config);
		this.config = config;
	}

	async solve(
		query: string,
		initialSolution?: string | undefined | null,
	): Promise<ReasoningTrace> {
		const maxTrials = this.config.maxTrials ?? 3;
		let attempt = initialSolution ?? "";
		const reflections: string[] = [];

		for (let _ = 0; _ < maxTrials; _++) {
			if (!attempt) {
				attempt = await this._chat([
					{
						role: "user",
						content: `${query}\nThink step by step. End with 'Final answer: ...'.`,
					},
				]);
			}

			const critique = await this._chat([
				{
					role: "user",
					content: `[Problem]\n${query}\n\n[Attempt]\n${attempt}\n\nCritique weaknesses or errors.`,
				},
			]);
			reflections.push(critique);

			attempt = await this._chat([
				{
					role: "user",
					content: [
						"[Problem]",
						query,
						"",
						"Reflections:",
						...reflections,
						"",
						"Rewrite solution based on reflections.",
						"End with 'Final answer: ...'.",
					].join("\n"),
				},
			]);
		}

		const [reasoning, answer] = this._split(attempt);
		return {
			reasoning,
			answer,
			metadata: { method: "reflexion", reflections },
		};
	}
}
